// @stoqey/timeout-manager/dist/interval/interval.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var IntervalStatus;
(function (IntervalStatus) {
    IntervalStatus[IntervalStatus["Running"] = 0] = "Running";
    IntervalStatus[IntervalStatus["Completed"] = 1] = "Completed";
})(IntervalStatus = exports.IntervalStatus || (exports.IntervalStatus = {}));
var IntervalModel = /** @class */ (function () {
    function IntervalModel(handler, interval) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        this.uuid = this._generateUuid();
        this.handler = handler;
        this.interval = interval;
        this.arguments = args;
        this.id = null;
        this.timestamp = Date.now();
        this.status = IntervalStatus.Running;
    }
    IntervalModel.prototype._generateUuid = function () {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
            var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    };
    return IntervalModel;
}());
exports.IntervalModel = IntervalModel;
