// @stoqey/ibkr/dist/events/IBKREVENTS.const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var IBKREVENTS;
(function (IBKREVENTS) {
    IBKREVENTS["READY"] = "READY";
    // Connections
    IBKREVENTS["PING"] = "ping";
    IBKREVENTS["DISCONNECTED"] = "disconnected";
    IBKREVENTS["CONNECTED"] = "connected";
    IBKREVENTS["ERROR"] = "error";
    /**
     * @interface IBKRAccountSummary
     */
    IBKREVENTS["ON_ACCOUNT_SUMMARY"] = "on_account_summary";
    // Orders portfolios, historicalData,
    // --------------------------------------------------------------------------------------
    // --------------------------------------------ORDERS------------------------------------
    /**
     * [OrderWithContract]
     */
    IBKREVENTS["OPEN_ORDERS"] = "open_orders";
    /**
     * OrderStock
     */
    IBKREVENTS["PLACE_ORDER"] = "place_order";
    /**
     * { orderStatus: OrderStatus, order: OrderWithContract }
     */
    IBKREVENTS["ORDER_STATUS"] = "order_status";
    /**
     * { sale: CreateSale, order: OrderWithContract }
     */
    IBKREVENTS["ORDER_FILLED"] = "order_filled";
    IBKREVENTS["CREATE_SALE"] = "create_sale";
    // --------------------------------------------------------------------------------------
    // ---------------------------------------------PORTFOLIOS-------------------------------
    /**
     * [PortFolioUpdate]
     */
    IBKREVENTS["PORTFOLIOS"] = "portfolios";
    // --------------------------------------------------------------------------------------
    // ---------------------------------------------MARKETDATA-------------------------------
    /**
   * @interface SymbolWithMarketData
   * {    symbol: string,
          marketData: HistoryData[]
     }
   */
    IBKREVENTS["ON_MARKET_DATA"] = "on_market_data";
    IBKREVENTS["GET_MARKET_DATA"] = "get_market_data";
    /**
     * { symbol, tickType?: 'ASK' | 'BID' | 'CLOSE' }
     */
    IBKREVENTS["SUBSCRIBE_PRICE_UPDATES"] = "subscribe_price_updates";
    /**
     * { symbol: string; date: Date; close: number; }
     */
    IBKREVENTS["ON_PRICE_UPDATES"] = "subscribe_price_updates";
})(IBKREVENTS = exports.IBKREVENTS || (exports.IBKREVENTS = {}));
